README.md
[![Build Status](https://travis-ci.org/mukuge/nalist.el.svg?branch=dev)](https://travis-ci.org/mukuge/nalist.el)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

# nalist.el

API to deal with association lists with names, or named alists. It's safe with lexical and buffer-local bindings.

## Installation

Place `nalist.el` on a directory in your `load-path', and add this to
your Emacs config:

```el
(require 'nalist)
```

## API

### Initializers

* [nalist-init](#nalist-init-name-alist-key-shallow) `(name alist &key shallow)`
* [nalist-copy](#nalist-copy-nalist-old-nalist-new-key-shallow) `(nalist-old nalist-new &key shallow)`
* [nalist-make-local-variable](#nalist-make-local-variable-nalist) `(nalist)`
* [nalist-make-variable-buffer-local](#nalist-make-variable-buffer-local-nalist) `(nalist)`

### Accessors

* [nalist-get](#nalist-get-key-nalist-key-default-testfn-eq) `(key nalist &key default (testfn 'eq))`
* [nalist-set](#nalist-set-key-value-nalist-key-testfn-eq) `(key value nalist &key (testfn 'eq))`
* [nalist-remove](#nalist-remove-key-nalist-key-testfn-eq) `(key nalist &key (testfn 'eq))`
* [nalist-clear](#nalist-clear-nalist) `(nalist)`
* [nalist-pop](#nalist-pop-key-nalist-key-testfn-eq) `(key nalist &key (testfn 'eq)`
* [nalist-poppair](#nalist-pop-key-nalist-key-testfn-eq) `(nalist)`
* [nalist-pairs](#nalist-pop-key-nalist-key-testfn-eq) `(nalist)`
* [nalist-keys](#nalist-poppair-nalist) `(nalist)`
* [nalist-values](#nalist-pairs-nalist) `(nalist)`

### Iterator

* [nalist-map](#nalist-map-function-nalist) `(function nalist)`

### Predicates

* [nalist-pairp](#nalist-pairp-obj) `(obj)`
* [nalist-proper-list-p](#nalist-poper-list-p-obj) `(obj)`
* [nalist-nalist-p](#nalist-nalist-p) `(obj)`
* [nalist-equal](#nalist-equal-nalist-a-nalist-b) `(nalist-a nalist-b)`
* [nalist-subset-p](#nalist-subset-p-nalist-a-nalist-b) `(nalist-a nalist-b)`
* [nalist-set-equal-p](#nalist-set-equal-p-nalist-a-nalist-b) `(nalist-a nalist-b)`

## Documentation and Examples

### nalist-init `(name alist &key shallow)`

Bind NAME to ALIST if SHALLOW is non-nil, otherwise to a deep-copy of ALIST.

(fn NAME ALIST &key SHALLOW)

```lisp
(nalist-init nal-1 '((a . b) (c . d))) ;; => nal-1
(nalist-init nal-2 '((1 . a) (1.0 . b))) ;; => nal-2
(nalist-init nal-3 '(("foo" . 3) ((a (b c)) . d))) ;; => nal-3

```

### nalist-get `(key nalist &key default (testfn 'eq))`

Return the value of KEY in NALIST if found with TESTFN, otherwise DEFAULT.

The key lookup is done with TESTFN if non-nil, otherwise with
‘eq’.

(fn KEY NALIST &key DEFAULT (TESTFN 'eq))

```lisp
(nalist-get 'c nal-1) ;; => 'd
(nalist-get 1.0 nal-2) ;; => nil
(nalist-get 1.0 nal-2 :testfn 'eql) ;; => 'b
(nalist-get "foo" nal-3 :testfn 'equal) ;; => 3
```

### nalist-set `(key value nalist &key (testfn 'eq))`

Find the pair with KEY in NALIST with TESTFN, and set its value to VALUE.

This macro destructively changes the value of the pair with KEY
into VALUE if the pair with KEY already exists, otherwise add a
new pair with KEY and VALUE to NALIST.

NALIST needs to be a symbol without a quote to access the correct
binding in its context.

It returns VALUE.

(fn KEY VALUE NALIST &key (TESTFN ''eq))

```lisp
(nalist-set 'c 'g nal-1) ;; => 'g
;; The content of nal-1 is changed to '((a . b) (c . g))
```

### nalist-remove `(key nalist &key (testfn 'eq))`

Remove the pair with KEY from NALIST if found with TESTFN.

(fn KEY NALIST &key (TESTFN ''eq))

```lisp
(nalist-remove 'c nal-1) ;; => nil
;; The content of nal-1 is changed to '((a . b))
```

### nalist-pop `(key nalist &key (testfn 'eq)`

Remove the pair with KEY from NALIST, and return the value of the pair.

This macro uses TESTFN to find the pair with the KEY. The default
value of TESTFN is ‘eq’.

This is an alias of ‘nalist-remove’.

```lisp
(nalist-pop 'c nal-1) ;; => 'd
;; The content of nal-1 is changed to '((a . b))
```

### nalist-poppair `(nalist)`

Return a pair in NALIST, and remove it from NALIST.

```lisp
(nalist-poppair nal-1) ;; => '(a . b)
;; The content of nal-1 is changed to '((c . d))
```

### nalist-pairs `(nalist)`

Return a list consisting all the pairs in NALIST.

```lisp
(nalist-pairs nal-1) ;; => '((a . b) (c . d))
```

### nalist-keys `(nalist)`

Return a list consisting all the keys in NALIST.

```lisp
(nalist-keys nal-1) ;; => '(a c)
```

### nalist-values `(nalist)`

Return a list consisting all the values in NALIST.

```lisp
(nalist-values nal-1) ;; => '(b d)
```

### nalist-copy `(nalist-old nalist-new &key shallow)`

Copy and bind the content of NALIST-OLD to NALIST-NEW.

This macro uses shallow-copy if SHALLOW is non-nil, otherwise
uses deep-copy.

(fn NALIST-OLD NALIST-NEW &key SHALLOW)

```lisp
(nalist-copy nal-1 nal-deep) ;; => 'nal-deep
(nalist-copy nal-1 nal-shallow :shallow t) ;; => 'nal-shallow
```

### nalist-clear `(nalist)`

Set NALIST nil.

```lisp
(nalist-clear nal-1) ;; => nil
```

### nalist-map `(function nalist)`

Call FUNCTION for all pairs in NALIST.

FUNCTION is called with two arguments, KEY and VALUE.
‘nalist-map’ always returns nil.

```lisp
(let ((res nil))
  (nalist-map #'(lambda (k v) (push k res)) nal-1)
  res) ;; => '(a c)
```

### nalist-pairp `(obj)`

Return t if OBJ is a pair, otherwise nil.

A pair is a cons cell, regardless of what Lisp objects its car
and cdr point.

```lisp
(nalist-pairp 'a) ;; => nil
(nalist-pairp '(a . b)) ;; => t
```

### nalist-proper-list-p `(obj)`

Return t if OBJ is a proper list, otherwise nil.

A proper list is a non circular cons chain whose last ‘cdr’ points nil.

```lisp
(nalist-proper-list-p nil) ;; => t
(nalist-proper-list-p '#1=(#1# . #1#)) ;; => nil
(nalist-proper-list-p '#1=(#1#  #1#)) ;; => t
(nalist-proper-list-p '(a)) ;; => t
(nalist-proper-list-p '(a . b)) ;; => nil

```

### nalist-nalist-p `(obj)`

Return t if OBJ is an (n)alist, otherwise nil.

An alist, or association list, is a proper list of pairs.  What
‘car’ and ‘cdr’ of a pair in alist point is often called a key
and value respectively.

```lisp
(nalist-nalist-p nil) ;; => t
(nalist-nalist-p '(a . b)) ;; => nil
(nalist-nalist-p '((a . b) . c)) ;; => nil
(nalist-nalist-p '((a . b) (c . d))) ;; => t
```

### nalist-subset-p `(nalist-a nalist-b)`

Return t if NALIST-A is a subset of NALIST-B with ‘equal’, otherwise nil.

```lisp
(nalist-subset-p '((a . b)) '((a . b) (c . d))) ;; => t
(nalist-subset-p '((a . b) (c . d)) '((a . b))) ;; => nil
```

### nalist-equal `(nalist-a nalist-b)`

Return t if NALIST-A nad NALIST-B are identical with ‘equal’, otherwise nil.

```lisp
(nalist-equal nal-1 '((a . b) (c . d)))) ;; => t
(nalist-equal nal-1 '((c . d) (a . b)))) ;; => nil
```

### nalist-set-equal-p `(nalist-a nalist-b)`

Test with ‘equal’ if NALIST-A and NALIST-B contain the same set of pairs.

Return t if so, otherwise nil.

```lisp
(nalist-set-equal nal-1 '((a . b) (c . d)))) ;; => t
(nalist-set-equal nal-1 '((c . d) (a . b)))) ;; => t
```

### nalist-make-local-variable `(nalist)`

Create a buffer-local binding in the current buffer for NALIST.

This macro binds a deep-copy of the content of the original
NALIST to the buffer-local NALIST to avoid their sharing cons
cells.

```lisp
(nalist-make-local-variable na-1)
;; na-1 is changed to a buffer-local variable in the current buffer.
```

### nalist-make-variable-buffer-local `(nalist)`

Mark NALIST automatically buffer local.

This macro binds a deep-copy of the content of the original
NALIST to the buffer-local NALIST to avoid their sharing cons
cells.

It also sets the default value of NALIST to nil to avoid the
buffer-local variables in other buffers share the cons cells
through it.

```lisp
(nalist-make-variable-buffer-local na-2)
;; na-2 is changed to an automatically buffer-local variable.
```


CHANGELOG.md
# Changelog

## 0.1.6.d (wip)

Preparing to upload to MELPA.

## 0.1.5 (2019-02-25)

### Changes

* Added API documentation.


## 0.1.4 (2019-02-19)

### Changes

* Added prefixes to all `cl-gensym` calls for slightly easier macro debugging.
* Dropped ALIST-EVAL-ONCE from `nalist-init` and make the feature default.


## 0.1.3 (2019-02-18)

### Changes

* Added a keyword argmunet ALIST-EVAL-ONCE to `nalist-init`.

* On Emacs 26, the return value of `nalist-set` was changed from the
  whole alist to the value just associated with the key.

### Bug fixes

* Fixed leaky macros: `nalist-set` and `nalist-remove`.


## 0.1.2 (2019-02-16)

### New features

* Implemented the testfn feature on Emacs 25 for `nalist-get`, `nalist-set`, and `nalist-remove`.

### Changes

* Changed the return value of `nalist-remove` to be the same as `nalist-pop`.
* Changed `nalist-pop` from a macro to an alias of `nalist-remove`.


## 0.1.1 (2019-02-16)

### Changes

* Deployed Travis CI.
* Added support for Emacs 25.1, 25.2, and 25.3.
* On Emacs 25, testfn of `nalist-get` is ignored and always 'eq.


test/nalist-init.el
;;; nalist-init.el --- Nalist: Initialization for the tests.  -*- lexical-binding: t; -*-
;; Copyright (C) 2019  Cyriakus "Mukuge" Hill

;; Author: Cyriakus "Mukuge" Hill <cyriakus.h@gmail.com>
;; Keywords: Lisp, tools
;; URL: https://github.com/mukuge/nalist.el

;; This file is NOT part of GNU Emacs.

;;; License:

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'seq)
(require 'ert)
(require 'f)
(require 's)

(defvar nalist-test/test-path
  (directory-file-name (file-name-directory load-file-name))
  "Path to tests directory.")

(defvar nalist-test/root-path
  (directory-file-name (file-name-directory nalist-test/test-path))
  "Path to root directory.")

(load (expand-file-name "nalist" nalist-test/root-path) 'noerror 'nomessage)

;;(load (f-expand "nalist.el" default-directory))

(provide 'nalist-init)
;;; nalist-init.el ends here


test/test-helper.el
;;; test-helper.el --- Nalist: Helpers for the tests. -*- lexical-binding: t; -*-
;; Copyright (C) 2019  Cyriakus "Mukuge" Hill

;; Author: Cyriakus "Mukuge" Hill <cyriakus.h@gmail.com>
;; Keywords: Lisp, tools
;; URL: https://github.com/mukuge/nalist.el

;; This file is NOT part of GNU Emacs.

;;; License:

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(defun nalist-helper-subset-p (nalist-a nalist-b)
  "Return t if NALIST-A is a subset of NALIST-B, otherwise nil.

Equality is defined by `equal'."
  (let ((res t))
    (mapc #'(lambda (pair)
              (unless (member pair nalist-b)
                (setq res nil)))
          nalist-a)
    res))

(defmacro nalist-helper-seq-set-equal-p (sequence1 sequence2 &optional testfn)
  "Return non-nil if SEQUENCE1 and SEQUENCE2 contain the same elements.

Equality is defined by `equal' in Emacs 25. In Emacs newer than
version 26, it is defined by TESTFN if non-nil or by `equal’ if
nil."
  (if (>= emacs-major-version 26)
      `(seq-set-equal-p ,sequence1 ,sequence2 ,testfn)
    `(and (nalist-helper-subset-p ,sequence1 ,sequence2)
          (nalist-helper-subset-p ,sequence2 ,sequence1))))

(defun nalist-helper-compose-generate-buffer-forms (sym-list)
  "Compose forms from SYM-LIST for `with-temp-buffers'."
  (mapcar #'(lambda (sym)
              `(setq ,sym (generate-new-buffer "temp")))
          sym-list))

(defun nalist-helper-compose-kill-buffer-forms (sym-list)
  "Compose forms from SYM-LIST for `with-temp-buffers'."
  (mapcar #'(lambda (sym)
              `(kill-buffer ,sym))
          sym-list))

(defmacro with-temp-buffers (sym-list &rest body)
  "Generate new buffers, bind them to symbols in SYM-LIST, and do BODY."
  (declare (indent 1))
  `(unwind-protect
       (progn
         ,@(nalist-helper-compose-generate-buffer-forms sym-list)
         ,@body)
     ,@(nalist-helper-compose-kill-buffer-forms sym-list)))

(defmacro should-assertion-error (&rest body)
  "Make `cl-assert' raise an error while executing BODY."
  (declare (indent 1))
  `(let ((debug-on-error nil))
     (should-error
      ,@body)))

(defun nalist-helper-compose-unbound-forms (sym-list)
  "Compose forms from SYM-LIST for `with-unbound-symbols'."
  (mapcar #'(lambda (sym)
              `(makunbound ',sym))
          sym-list))

(defmacro with-unbound-symbols (sym-list &rest body)
  "Unbound all the vars in SYM-LIST before executing BODY."
  (declare (indent 1))
  `(progn
     ,@(nalist-helper-compose-unbound-forms sym-list)
     ,@body))

(defun nalist-helper-compose-forms (var-list)
  "Compose forms from VAR-LIST for `with-temp-conses'."
  (mapcar #'(lambda (var)
              `(nalist-helper-totally-destruct-cons ,var))
          var-list))

(defmacro with-temp-conses (var-list &rest body)
  "Destruct all conses the variables in VAR-LIST bound, after executing BODY."
  (declare (indent 1))
  `(unwind-protect
       (progn
         ,@body)
     ,@(nalist-helper-compose-forms var-list)))

(defmacro with-nalist-fixture (&rest body)
  `(unwind-protect
       (progn
         ;; setup
         (nalist-init nal '((a . b) (c . d)))
         (nalist-init nal-eql '((1 . a) (1.0 . b)))
         (nalist-init nal-equal '(("foo" . 3) ((a (b c)) . d)))
         (nalist-init nal-4 '((a . b) (c . d) (e . f) (g . h)))
         ,@body)
     ;; teardown
     ))

(defmacro nalist-helper-totally-destruct-cons (obj)
  "Totally destruct OBJ.

OBJ should be a cons cell or a symbol refers a con cell."
  `(prog1
       (unless (atom ,obj)
         (nalist-helper-chop-all-leaves ,obj))
     (when (symbolp ',obj)
       (setq ,obj nil))
     (garbage-collect)))

(defun nalist-helper-chop-all-leaves (cons-cell)
  "Dereference all cells of CONS-CELL from tips to the root."
  (unless (or (null cons-cell)
              (equal cons-cell '(nil . nil)))
    (when (nalist-helper-car-has-leaves cons-cell)
      (nalist-helper-chop-all-leaves (car cons-cell)))
    (when (nalist-helper-cdr-has-leaves cons-cell)
      (nalist-helper-chop-all-leaves (cdr cons-cell)))
    (nalist-helper-chop-all-leaves cons-cell)))

(defun nalist-helper-cdr-has-leaves (cons-cell)
  "Return t if the cdr of CONS-CELL points more than an atom or '(nil).

If the cdr poinsts nil, just return nil.  If the cdr points a
non-nil atom or '(nil), `setcdr' nil and return nil.  Otherwise,
return t."
  (cond ((null (cdr cons-cell)))
        ((atom (cdr cons-cell))
         (setcdr cons-cell nil))
        ((and (null (cdar cons-cell))
              (null (cddr cons-cell)))
         (setcdr cons-cell nil))
        (t t)))

(defun nalist-helper-car-has-leaves (cons-cell)
  "Return t if the car of CONS-CELL points more than an atom or '(nil).

If the car points nil, just return nil.  If the car points a
non-nil atom or '(nil), `setcar' nil and return nil.  Otherwise,
return t."
  (cond ((null (car cons-cell)))
        ((atom (car cons-cell))
         (setcar cons-cell nil))
        ((and (null (caar cons-cell))
              (null (cadr cons-cell)))
         (setcar cons-cell nil))
        (t t)))

(defun nalist-helper-destruct-cons (cons-cell)
  "Destruct CONS-CELL completely."
  (cond ((null cons-cell))
        ;; Fisrtly, we deal with the cases one of the item in the
        ;; considerdation is a non-nil atom.  Because applying
        ;; cons-related functions to non-nil atoms causes erros.
        ((and (atom (car cons-cell))
              (atom (cdr cons-cell)))
         (setcar cons-cell nil)
         (setcdr cons-cell nil)
         ;; In case one of these is nil, it's redundant to set it to
         ;; nil. However, it's more efficient to allow it than adding
         ;; conditional branched for it since this is an end of the
         ;; iteration.
         )
        ((atom (car cons-cell))
         (setcar cons-cell nil) ;; This is redundant, when car is nil, but not harmful.
         (if (and (null (cdar cons-cell))
                  (null (cddr cons-cell)))
             (setcdr cons-cell nil)
           (nalist-helper-destruct-cons (cdr cons-cell)))
         (nalist-helper-destruct-cons cons-cell))
        ((atom (cdr cons-cell))
         (setcdr cons-cell nil) ;; This is redundant, when car is nil, but not harmful.
         (if (and (null (caar cons-cell))
                  (null (cadr cons-cell)))
             (setcar cons-cell nil)
           (nalist-helper-destruct-cons (car cons-cell)))
         (nalist-helper-destruct-cons cons-cell))
        ;; Now both car and cdr are consp.
        ((and (null (caar cons-cell))
              (null (cadr cons-cell))
              (null (cdar cons-cell))
              (null (cddr cons-cell)))
         (setcar cons-cell nil)
         (setcdr cons-cell nil))
        ((and (null (caar cons-cell))
              (null (cadr cons-cell)))
         (setcar cons-cell nil)
         (nalist-helper-destruct-cons (cdr cons-cell))
         (nalist-helper-destruct-cons cons-cell))
        ((and (null (cdar cons-cell))
              (null (cddr cons-cell)))
         (setcdr cons-cell nil)
         (nalist-helper-destruct-cons (car cons-cell))
         (nalist-helper-destruct-cons cons-cell))
        (t (nalist-helper-destruct-cons (car cons-cell))
           (nalist-helper-destruct-cons (cdr cons-cell))
           (nalist-helper-destruct-cons cons-cell))))

;;; test-helper.el ends here


nalist.el
;;; nalist.el --- API for named association lists.  -*- lexical-binding: t; -*-

;; Copyright (C) 2019  Cyriakus "Mukuge" Hill

;; Author: Cyriakus "Mukuge" Hill <cyriakus.h@gmail.com>
;; Keywords: Lisp, tools
;; URL: https://github.com/mukuge/nalist.el
;; Package-Version: 0.1.6.d
;; Package-Requires: ((emacs "25.1"))

;; This file is NOT part of GNU Emacs.

;;; License:

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program. If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;; API for named association lists.

;;; Code:

(require 'cl-lib)
(require 'seq)

(defun nalist-pairp (obj)
  "Return t if OBJ is a pair, otherwise nil.

A pair is a cons cell, regardless of what Lisp objects its car
and cdr point."
  (consp obj))

(defun nalist-proper-list-p (obj)
  "Return t if OBJ is a proper list, otherwise nil.

A proper list is a non circular cons chain whose last `cdr' points nil."
  (or
   (null obj)
   (and (consp obj)
        (cl-loop
         for nth-cdr = obj then (cdr nth-cdr)
         for 2nth-cdr = (cdr obj) then (cddr 2nth-cdr)
         when (eq nth-cdr 2nth-cdr) return nil
         ;; A circular list consists of finite positions: When we move
         ;; a pointer along with its cons cell chain by using `cdr',
         ;; it goes back to the start point in finite steps.
         ;; Moreover, it's periodic.  Let's assume its period is N
         ;; steps.  If we have two pointers so that one moves one step
         ;; at a time starting from the position 1, and the other
         ;; moves two steps at a time starting from the position 2,
         ;; the former points the nth position, and the latter points
         ;; the 2nth position after n iterations.  So, after N
         ;; iterations, the former and the latter point the same
         ;; position on the circular list because it's N-periodic.
         ;;
         when (null nth-cdr) return t
         ;; nth-cdr and 2nth-cdr advance in different steps at a
         ;; time. It means there are some points which the nth-cdr
         ;; visits before the 2nth-cdr.  So, we need to check nullness
         ;; of not only 2nth-cdr, the faster runner, but also the
         ;; nth-cdr.
         ;;
         when (null 2nth-cdr) return t
         when (not (consp 2nth-cdr)) return nil
         ;; If 2nth cdr position is a terminus, it's a non circular
         ;; list.  If the terminous is nil, it's a proper list.  If
         ;; the termious is also not consp, i.e. it's an atom other
         ;; than nil, it's not a proper list.
         ;;
         when (null (cdr 2nth-cdr)) return t
         when (not (consp (cdr 2nth-cdr))) return nil
         ;; If (2n+1)th cdr position is a terminus, it's a non
         ;; circular list.  If the terminous is nil, it's a proper
         ;; list. If the termious is also not consp, i.e. it's an atom
         ;; other than nil, it's not a proper list.
         ;;
         when (null (cddr 2nth-cdr)) return t
         when (not (consp (cddr 2nth-cdr))) return nil)
        ;; If (2n+2)th, or the next 2nth, cdr position is a terminous,
        ;; it's a non circular list.  If the terminous is nil, it's a
        ;; proper list, and if the termious is also not consp,
        ;; i.e. it's an atom other than nil, it's not a proper list.
        ;;
        ;; Given none of the three positions above is a terminus,
        ;; iterate over to the next nth and 2nth, i.e. (n+1)th and
        ;; 2(n+1)th, positions, which are nicely expressed with the
        ;; `cdr' for nth-cdr and `cddr' for 2nth-cdr in the variable
        ;; definition part of this `cl-loop'.
        )))

(defun nalist-nalist-p (obj)
  "Return t if OBJ is an (n)alist, otherwise nil.

An alist, or association list, is a proper list of pairs.  What
`car' and `cdr' of a pair in alist point is often called a key
and value respectively."
  (and (nalist-proper-list-p obj)
       (let ((res t))
         (mapc #'(lambda (pair)
                   (unless (nalist-pairp pair)
                     (setq res nil)))
               obj)
         res)))

(cl-defmacro nalist-init (name alist &key shallow)
  "Bind NAME to ALIST if SHALLOW is non-nil, otherwise to a deep-copy of ALIST."
  (let ((ealist (cl-gensym "alist-")))
    `(let ((,ealist ,alist))
       (unless (nalist-nalist-p ,ealist)
         (error "Invalid initial value `%s'" ,ealist))
       (if ,shallow
           (setq ,name ,ealist)
         (setq ,name (copy-alist ,ealist)))
       ',name)))

(defmacro nalist-make-local-variable (nalist)
  "Create a buffer-local binding in the current buffer for NALIST.

This macro binds a deep-copy of the content of the original
NALIST to the buffer-local NALIST to avoid their sharing cons
cells."
  (let ((copy (cl-gensym "copy-")))
    `(let ((,copy (copy-alist ,nalist)))
       (make-local-variable ',nalist)
       (setq ,nalist ,copy)
       ',nalist)))

(defmacro nalist-make-variable-buffer-local (nalist)
  "Mark NALIST automatically buffer local.

This macro binds a deep-copy of the content of the original
NALIST to the buffer-local NALIST to avoid their sharing cons
cells.

It also sets the default value of NALIST to nil to avoid the
buffer-local variables in other buffers share the cons cells
through it."
  (let ((copy (cl-gensym "copy-")))
    `(let ((,copy ,nalist))
       (setq ,nalist nil)
       (make-variable-buffer-local ',nalist)
       (setq ,nalist ,copy)
       ',nalist)))

(defmacro nalist-clear (nalist)
  "Set NALIST nil."
  `(setq ,nalist nil))

(cl-defmacro nalist--alist-set (key value nalist &key (testfn ''eq))
  "Find the pair with KEY in NALIST with TESTFN, and set its value to VALUE.

This is a compatibility macro for Emacs 25."
  (let ((before (cl-gensym "before-"))
        (after (cl-gensym "after-"))
        (pair (cl-gensym "pair-"))
        (pair-found (cl-gensym "pair-found-"))
        (ekey (cl-gensym "key-"))
        (evalue (cl-gensym "value-")))
    `(progn
       (cl-assert (nalist-nalist-p ,nalist) t)
       (let ((,ekey ,key)
             (,evalue ,value))
         (if (null ,nalist)
             (progn
               (setq ,nalist (list (cons ,ekey ,evalue)))
               ,value)
           (cl-do* ((,before nil)
                    (,after ,nalist (cdr ,after))
                    (,pair (car ,after) (car ,after))
                    (,pair-found nil))
               ((null ,after)
                (unless ,pair-found
                  (push (cons ,ekey ,evalue) ,before))
                (setq ,nalist ,before)
                ,evalue)
             (when (funcall ,testfn (car ,pair) ,ekey)
               (setq ,pair-found t)
               (setq ,pair (cons (car ,pair) ,evalue)))
             (push ,pair ,before)))))))

(cl-defmacro nalist-set (key value nalist &key (testfn ''eq))
  "Find the pair with KEY in NALIST with TESTFN, and set its value to VALUE.

This macro destructively changes the value of the pair with KEY
into VALUE if the pair with KEY already exists, otherwise add a
new pair with KEY and VALUE to NALIST.

NALIST needs to be a symbol without a quote to access the correct
binding in its context.

It returns VALUE."
  (let ((evalue (cl-gensym "value-")))
    (if (>= emacs-major-version 26)
        `(let ((,evalue ,value))
           (setf (alist-get ,key ,nalist nil nil ,testfn) ,evalue)
           ,evalue)
      `(let ((,evalue ,value))
         (nalist--alist-set ,key ,evalue ,nalist :testfn ,testfn)))))

(cl-defun nalist--alist-get (key nalist &key (default nil) (testfn 'eq))
  "Return the value associated with KEY in ALIST with using TESTFN.

This is a compatibility function for Emacs 25."
  (let ((list nalist)
        (res default))
    (while list
      (let ((pair (car list)))
        (when (funcall testfn key (car pair))
          (setq res (cdr pair))))
      (setq list (cdr list)))
    res))

(cl-defun nalist-get (key nalist &key default (testfn 'eq))
  "Return the value of KEY in NALIST if found with TESTFN, otherwise DEFAULT.

The key lookup is done with TESTFN if non-nil, otherwise with
`eq'."
  (if (>= emacs-major-version 26)
      (alist-get key nalist default nil testfn)
    (nalist--alist-get key nalist :default default :testfn testfn)))

(cl-defmacro nalist--remove (key nalist &key (default nil) (testfn ''eq))
  "Remove the pair with KEY from NALIST, and return the value of the pair.

This macro uses TESTFN to find the pair with the KEY. The default
value of TESTFN is `eq'.

This is a compatibility function for Emacs 25."
  (let ((before (cl-gensym "before-"))
        (after (cl-gensym "after-"))
        (pair (cl-gensym "pair-"))
        (pair-found (cl-gensym "pair-found-")))
    `(progn
       (cl-assert (nalist-nalist-p ,nalist) t)
       (if (null ,nalist) nil
         (cl-do* ((,before nil)
                  (,after ,nalist (cdr ,after))
                  (,pair (car ,after) (car ,after))
                  (,pair-found nil))
             ((null ,after)
              (setq ,nalist ,before)
              (if ,pair-found
                  (cdr ,pair-found)
                ,default))
           (if (funcall ,testfn (car ,pair) ,key)
               (setq ,pair-found ,pair)
             (push ,pair ,before)))))))

(cl-defmacro nalist-remove (key nalist &key (testfn ''eq))
  "Remove the pair with KEY from NALIST if found with TESTFN."
  (if (>= emacs-major-version 26)
      (let ((ekey (cl-gensym "key-"))
            (etestfn (cl-gensym "testfn-"))
            (value (cl-gensym "value-")))
        `(let ((,ekey ,key)
               (,etestfn ,testfn))
           (let ((,value (alist-get ,ekey ,nalist nil t ,etestfn)))
             (setf (alist-get ,ekey ,nalist nil t ,etestfn) nil)
             ,value)))
    `(nalist--remove ,key ,nalist :default nil :testfn ,testfn)))

(defun nalist-pairs (nalist)
  "Return a list consisting all the pairs in NALIST."
  (cl-assert (nalist-nalist-p nalist) t)
  (copy-alist nalist))

(defun nalist-keys (nalist)
  "Return a list consisting all the keys in NALIST."
  (cl-assert (nalist-nalist-p nalist) t)
  (mapcar 'car nalist))

(defun nalist-values (nalist)
  "Return a list consisting all the values in NALIST."
  (cl-assert (nalist-nalist-p nalist) t)
  (mapcar 'cdr nalist))

(cl-defmacro nalist-copy (nalist-old nalist-new &key shallow)
  "Copy and bind the content of NALIST-OLD to NALIST-NEW.

This macro uses shallow-copy if SHALLOW is non-nil, otherwise
uses deep-copy."
  `(progn
     (cl-assert (nalist-nalist-p ,nalist-old) t)
     (if ,shallow
         (setq ,nalist-new ,nalist-old)
       (setq ,nalist-new (copy-alist ,nalist-old)))
     ',nalist-new))

(defalias 'nalist-pop 'nalist-remove
  "Remove the pair with KEY from NALIST, and return the value of the pair.

This macro uses TESTFN to find the pair with the KEY. The default
value of TESTFN is `eq'.

This is an alias of `nalist-remove'.")

(defmacro nalist-poppair (nalist)
  "Return a pair in NALIST, and remove it from NALIST."
  (let ((pair (cl-gensym "pair-")))
    `(progn
       (cl-assert (nalist-nalist-p ,nalist))
       (let ((,pair (car ,nalist)))
         (setq ,nalist (cdr ,nalist))
         ,pair))))

(defun nalist-map (function nalist)
  "Call FUNCTION for all pairs in NALIST.

FUNCTION is called with two arguments, KEY and VALUE.
‘nalist-map’ always returns nil."
  (cl-assert (nalist-nalist-p nalist) t)
  (let ((remaining nalist))
    (while remaining
      (let ((pair (car remaining)))
        (funcall function (car pair) (cdr pair))
        (setq remaining (cdr remaining))))
    nil))

(defun nalist-subset-p (nalist-a nalist-b)
  "Return t if NALIST-A is a subset of NALIST-B with `equal', otherwise nil."
  (cl-assert (nalist-nalist-p nalist-a) t)
  (cl-assert (nalist-nalist-p nalist-b) t)
  (let ((res t))
    (mapc #'(lambda (pair)
              (unless (member pair nalist-b)
                (setq res nil)))
          nalist-a)
    res))

(defun nalist-equal (nalist-a nalist-b)
  "Return t if NALIST-A nad NALIST-B are identical with `equal', otherwise nil."
  (cl-assert (nalist-nalist-p nalist-a) t)
  (cl-assert (nalist-nalist-p nalist-b) t)
  (equal nalist-a nalist-b))

(defmacro nalist--compat-seq-set-equal-p (sequence1 sequence2 &optional testfn)
  "Return non-nil if SEQUENCE1 and SEQUENCE2 contain the same set of elements.

Equality is defined by `equal' on Emacs 25. On Emacs newer than
version 26, it is defined by TESTFN if non-nil or by `equal’ if
nil."
  (if (>= emacs-major-version 26)
      `(seq-set-equal-p ,sequence1 ,sequence2 ,testfn)
    `(and (nalist-subset-p ,sequence1 ,sequence2)
          (nalist-subset-p ,sequence2 ,sequence1))))

(defun nalist-set-equal-p (nalist-a nalist-b)
  "Test with `equal' if NALIST-A and NALIST-B contain the same set of pairs.

Return t if so, otherwise nil."
  (cl-assert (nalist-nalist-p nalist-a) t)
  (cl-assert (nalist-nalist-p nalist-b) t)
  (nalist--compat-seq-set-equal-p nalist-a nalist-b))



(provide 'nalist)
;;; nalist.el ends here
